### ISSUE TITLE
Hosting: Issues with StaticWebAssets in default builder

### ISSUE THREAD

**Issue by dazinator** [dotnet/aspnetcore#34890]
## Background and Motivation
Using the out of the box `CreateDefaultBuilder` style api's I end up with a web host that doesn't deal with `StaticWebAssets` the way i'd like.

Sadly as the [ConfigureWebDefaults](https://github.com/dotnet/aspnetcore/blob/18a926850f7374248e687ee64390e7f10514403f/src/DefaultBuilder/src/WebHost.cs#L215) method where some static assets logic is defined, is internal, I've had to copy it with the only change being a tweak to the static assets reader to better cater for my application logic around static apps deployments. 

https://github.com/dotnet/aspnetcore/blob/18a926850f7374248e687ee64390e7f10514403f/src/DefaultBuilder/src/WebHost.cs#L221

This exercise bought me to the issue at hand..  it seems that the startup filters / middlewares that are kept internal here are very generic and useful to any web host. I know i'm building a custom web host at this point, but its a shame I have to copy and paste these internal types to add forwarded headers and host filtering. If they were public it would be less copy paste on my side:
https://github.com/dotnet/aspnetcore/blob/18a926850f7374248e687ee64390e7f10514403f/src/DefaultBuilder/src/WebHost.cs#L231

## Proposed API
As per above, make the internal types for `HostFiltering` and `ForwardedHeaders`, public or provide extension methods to easily include these internal services when building a custom web host.

## Usage Examples

Perhaps provide public extension methods so these services can be added?

```
IWebHostBuilder builder = WillHaveThisWhenConfiguringWebHost()
builder.ConfigureServices((hostingContext, services) =>
            {
              // i.e these new extension methods so I don't have to copy internal types.
               services.AddForwardedHeaders()
                            .AddHostFiltering();            
            }
```

## Alternative Designs

Perhaps consider some refactor or change to `ConfigureWebDefaults` so that there is a mechanism to use it to still to configure a web host, but there are more options around taking control of static assets, or excluding the default static assets stuff that the framework currently does.

## Risks

If extension methods are provided they might light up for consumers who might call them to add the services a second time for example in their startup classes, by accident. Previoulsy as there are no extension methods and the types were internal, there would be no risk of this. This can be mitigated by ensuring the extension methods do `TryAdd` when adding the services to the service collection to avoid possibility of dupe registrations.

**Comment by Tratcher** [comment]
Note that Host filtering and forwarded headers are public middleware, the only internal component here is the IStartupFilters used to inject them into apps on your behalf. Since you're already working in your own app, you can call UseHostFiltering and UseForwardedHeaders directly.

https://github.com/dotnet/aspnetcore/blob/18a926850f7374248e687ee64390e7f10514403f/src/Middleware/HostFiltering/src/HostFilteringBuilderExtensions.cs#L20
https://github.com/dotnet/aspnetcore/blob/18a926850f7374248e687ee64390e7f10514403f/src/Middleware/HttpOverrides/src/ForwardedHeadersExtensions.cs#L26

> Perhaps consider some refactor or change to ConfigureWebDefaults so that there is a mechanism to use it to still to configure a web host, but there are more options around taking control of static assets, or excluding the default static assets stuff that the framework currently does.

Yes, making that more configurable is a good idea.

**Comment by dazinator** [comment]
> Since you're already working in your own app, you can call UseHostFiltering and UseForwardedHeaders directly.

Just to elaborate, my custom host is distributed as a nuget package. Users have their own Program Main and call an extension method that I provide on IHostBuilder to build this customised Web host as opposed to the native one (the only difference being a tweak to static assets in my implementation). They already have startup classes and startup code that is in place. In this instance they can't just swap in my custom host without also adding Forwarded Headers and HostFiltering themselves - unless I duplicate the internal startup filters so I can include it for them im my host which is what I've done. 

Not a big deal by any means but wanted you to have the full picture

**Comment by Tratcher** [comment]
> Just to elaborate, my extension method for building the webhost in a different way, is distributed as a nuget package.

Makes sense.  Given the infrastructure nature of these IStartupFilter types we don't think they're useful to most people and don't plan to make them public.

> Perhaps consider some refactor or change to ConfigureWebDefaults so that there is a mechanism to use it to still to configure a web host, but there are more options around taking control of static assets, or excluding the default static assets stuff that the framework currently does.

Can you elaborate on what issues you had with the current static assets logic? How would you like it to work instead?

**Comment by dazinator** [comment]
> > Just to elaborate, my extension method for building the webhost in a different way, is distributed as a nuget package.
> 
> Makes sense. Given the infrastructure nature of these IStartupFilter types we don't think they're useful to most people and don't plan to make them public.
> 
> > Perhaps consider some refactor or change to ConfigureWebDefaults so that there is a mechanism to use it to still to configure a web host, but there are more options around taking control of static assets, or excluding the default static assets stuff that the framework currently does.
> 
> Can you elaborate on what issues you had with the current static assets logic? How would you like it to work instead?

Sure!

Its a very small tweak but conceptually important for me. When running in development, the default Web host sets up file providers for each static asset base pathup checks for a static assets manifest file, which contains entries for any referenced blazor wasm spa apps. It then proceeds to look at

**Comment by dazinator** [comment]
> > Just to elaborate, my extension method for building the webhost in a different way, is distributed as a nuget package.
> 
> Makes sense. Given the infrastructure nature of these IStartupFilter types we don't think they're useful to most people and don't plan to make them public.
> 
> > Perhaps consider some refactor or change to ConfigureWebDefaults so that there is a mechanism to use it to still to configure a web host, but there are more options around taking control of static assets, or excluding the default static assets stuff that the framework currently does.
> 
> Can you elaborate on what issues you had with the current static assets logic? How would you like it to work instead?

Sure I can elaborate! I made a very small one line tweak but the rational behind it takes some explaining :-)

When running in development, the default Web host sets up independent file providers for each "static asset base path" in the manifest, to serve the physical content from each referenced  location. When you dotnet publish and run on a prod server, it instead relies on the content being copied under the webroot dir (in a subfolder named the same as the static asset base path) and the hosts singular / default webroot file provider is entrusted with serving content from those subfolders. The issues I have with this are:

1. Suppose your "static assets base path" starts with a "." e.g ".private/spa1". This will work in development but 404 on prod as the file provider on prod can't see subfolders beginning with "." but the individual fileproviders in dev point at physical locations without any "." and so it all works in dev. (Will cover my desire to use a "." in a mo). INIn next point) This strikes me as inconsistent behaviour between dev and prod and is a smell imho. 
2. The default web host assumes that in development, all referenced static assets (that appear in the manifest) should be served. However in prod, thanks to there being no manifest, only the stuff that's copied under wwwroot and visible to the physical file provider, will be servable by default. This means if you do use a ".private/spa1" as a static asset base path, the spa becomes `private` in production and will not be served up by the host by default. This ability to make spa's "private" / not servable by default is actually really powerful. It means I can now take control again over which spa's are servable and when. The middleware pipeline is a powerful thing that gives full control to developers over how requests are processed and what happens. The default webhost appears to take this control away for blazor spa's in terms of making the decision for you to serve up all blazor spa's for you on the paths in the manifest. As blazor wasm apps are just bunches of static files I'd rather have middleware that I can use to serve up individual blazor spa's in my solution at the point of my choosing. This could be after I have done Authorization checks, or more complex cases like:
- if the status of the app is in setup mode, I'll serve the "setup" spa on the root path /. If its not I'll serve the normal spa on the root path. The setup spa is only servable when the app is in setup mode it's not servable any other time.

It make sense to me that I'm able to leverage the middleware pipeline and chose when or if I'll serve a particular blazor spa.

So the change I made was very small, to make dev consistent with prod. If the static asset base path in the manifest contains a "." I ignore it and don't create a file provider. This matches how prod behaves with not seeing directories starting with a ".". To fully realise the vision however I then had to create custom middleware to allow me to serve up a given blazor spa from such a private location. The blazor middleware should exist for this out of the box imho

**Comment by dazinator** [comment]
Note: turns out I already logged an issue with respect to the "." stuff: https://github.com/dotnet/aspnetcore/issues/20605

**Comment by Tratcher** [comment]
Should this be closed as a duplicate of #20605?

**Comment by javiercn** [comment]
> 1. Suppose your "static assets base path" starts with a "." e.g ".private/spa1". This will work in development but 404 on prod as the file provider on prod can't see subfolders beginning with "." but the individual fileproviders in dev point at physical locations without any "." and so it all works in dev. (Will cover my desire to use a "." in a mo). This strikes me as inconsistent behaviour between dev and prod and is a smell imho.

I tried this out and this is not my experience with the file provider
![image](https://user-images.githubusercontent.com/6995051/128505839-6863a65c-1fbc-4ba2-af76-5328cd513316.png)

> 2\. The default web host assumes that in development, all referenced static assets (that appear in the manifest) should be served.

All assets in the manifest should be served, you can write MSBuild to add and remove assets. You can take control over the assets at build time and decide where to copy them or what to do with them.

I don't think that Static web assets is the right layer to do what you are trying to achieve. You would be best served with using  a custom piece of middleware to do the filtering or rewrite the path rather than trying to hijack the file system to do what you are trying to achieve.

The runtime support for static web assets is very scoped and limited to exposing the set of provided files as if they were part of the web root (where they'll originally end).

For the sake of completeness I also tried on a stock project template and the behavior is the same, the asset is served even when the path contains a `.` using the regular file provider.
![image](https://user-images.githubusercontent.com/6995051/128507120-fe678a0a-6edb-4f1a-b257-2a3041b8bebe.png)

Given that this is the case, I'm going to close this issue since static web assets is working as expected here.

**Comment by dazinator** [comment]
@javiercn 
> I tried this out and this is not my experience with the file provider

That's very odd. When I raised the original issue, it was certainly the case that "." prefixed directories published under the `wwwroot` (webroot base path) on production, were not visibile. This is also mystifying because:

1. You are constructing a PhysicalFileProvider with the `sensitive` [flag](https://docs.microsoft.com/en-us/dotnet/api/microsoft.extensions.fileproviders.physical.exclusionfilters?view=dotnet-plat-ext-5.0) which states that:

> Exclude files and directories when the name begins with a period, or has either Hidden or System is set on Attributes.

Therefore, this file provider by design should be excluding files and directories that with a ".". So how is it that your test reveals that they exist? I suspect its because you are using the full "hidden" path to resolve the directory or file. Could it be the case that static files middleware navigates the directory to the resource one path at a time? For example if you did this:

```
var directory = fp.GetDirectoryContents("/");

```

You would discover that the directory that is returned has no entry for the `.private' folder because this is hidden. If static files resolved whatever path was entered in the browser then files and directories that are hidden would be resolvable based on what you've shown, which is a MASSIVE security risk right and defeats the purpose of the flag "sensitive"

**Comment by javiercn** [comment]
> Therefore, this file provider by design should be excluding files and directories that with a ".". So how is it that your test reveals that they exist? I suspect its because you are using the full "hidden" path to resolve the directory or file. Could it be the case that static files middleware navigates the directory to the resource one path at a time? For example if you did this:

No, that's not the case. Those flags only filter files that start with `.` not anything on the intermediate path. If you were seeing something different in production it might have been some behavior IIS might have, but not something ASP.NET Core has.

Static files passes the path along down to the provider

**Comment by dazinator** [comment]
@javiercn 

> Those flags only filter files that start with . not anything on the intermediate path. 

From the [description](https://docs.microsoft.com/en-us/dotnet/api/microsoft.extensions.fileproviders.physical.exclusionfilters?view=dotnet-plat-ext-5.0) of the flag:

> Equivalent to DotPrefixed | Hidden | System. Exclude files and directories when the name begins with a period, or has either Hidden or System is set on Attributes.

It states that files **and directories** will be excluded. I'm still not understanding how you are able to navigate into directories that should supposedly be excluded? Perhaps there was some behaviour change from when I last looked at this but at that time, i'm 100% positive that the physical file provider looking at base wwwroot folder on a production server, could not serve content if they were in subfolders that were dot prefixed. If this has changed and static files worked as you descrive, then it means as long as you know the full path, dot prefixed directories are not in fact excluded, in which case the description of the enum probably needs updating? 

Basically I am not sure what it means to say dot prefixed directories are "excluded", but then be able to resolve them and serve content from them through the file provider. I note you didn't test with static files in the equation, the full path you'd have to send in the browser would have to include a "." i.e "/.private/spa1" - maybe static files has some cleansing of things like "."'s from the path in the request so it can never actually send through such a path to the underlying FP.

**Comment by dazinator** [comment]
P.S I should bever say 100% positive, i'm only human after all ;-)

**Comment by javiercn** [comment]
> Basically I am not sure what it means to say dot prefixed directories are "excluded", but then be able to resolve them and serve content from them through the file provider. I note you didn't test with static files in the equation, the full path you'd have to send in the browser would have to include a "." i.e "/.private/spa1" - maybe static files has some cleansing of things like "."'s from the path in the request so it can never actually send through such a path to the underlying FP.

I did test with static files in the mix, the second image is with a stock web app with a file in that path.

The file provider comments might be misleading, I don't think the behavior there changed since the creation, and probably the directory mention was an oversight. GetFileInfo won't return directories and `GetDirectoryContents` returns the contents whether or not the directory is prefixed with `.`. I imagine that the comment refers to files and results of enumerating directory contents.

**Comment by dazinator** [comment]
> I did test with static files in the mix, the second image is with a stock web app with a file in that path.

The second image, if its a stock web app running on a `Developement` environment, and you are using out of the box static assets stuff, will be demonstrating my point, because the out of the box static assets stuff on a Development environment does serve "hidden" content, it was only when published to prod that I was seeing hidden content no longer being served..

**Comment by javiercn** [comment]
It's not using static web assets, the file is on the wwwroot folder of the only project, which static web assets doesn't deal with

**Comment by dazinator** [comment]
> If you dotnet publish that same application, then run it on a non "Development" environment (i.e change the environment variable to Production) - I was seeing that browsing to that same path would 404.

@javiercn I can't repro this with `dotnet 5.0`, so apologies, I must have made a mistake somewhere or made a false assumption, or there was a behavior change since the problem exhibited, I was using `dotnet 3.X` at the time I beleive.

> The file provider comments might be misleading, I don't think the behavior there changed since the creation, and probably the directory mention was an oversight.

Yes I agree the comments seem to be a little misleading it would be good to clairfy what excluding actually means. You've shown that sensitive directories seem to exclude nothing!

Thanks for taking the time to look at this.

**Comment by dazinator** [comment]
Note in case someone else stumbles across this. Given out of the box file provider doesn't seem to exclude files in dot prefixed directories from being served by static files, if you did want more granular control over which diretories were visible to a file provider, you could use my (GlobPatternFilterFileProvider)[https://github.com/dazinator/Dazinator.Extensions.FileProviders#globpatternfilterfileprovider].
